docs/tendermint-core/p2p/peer_manager.md

```diff
@@ -0,0 +1,439 @@
+
+If the peer is still present in the peer store, its `LastDisconnected` time is
+set and the peer manager spawns a routine that after `DisconnectCooldownPeriod`
+notifies the next peer to dial routine about the availability of this peer.
+Until then, the peer is the `Frozen Candidate` state.
+
+### Errored transition
+
+This transition is performed when a reactor interacting with the peer reports
+an error to the router.
+
+The peer is expected to be in the [`Connected`](#connected-peer) state.
+If so, the peer transitions to the [`Evict`](#evict-peer) sub-state, which
+should lead the router to disconnect from the peer, and the next peer to evict
+routine is notified.
+
+### Upgrading peer
+
+A [connected peer](#connected-peer) which should be evicted to given room to a
+higher-ranked peer the router is dialing to.
+
+The `Upgrading` sub-state is part of the procedure to [upgrade connection slots](#slots-upgrade).
+When a connection with the higher-ranker peer that should take the connection
+slot from this peer is [established](#dialed-transition), the
+[eviction](#evict-peer) of this peer is scheduled.
+
+### Evict peer
+
+A peer whose eviction was scheduled, for either of the following reasons:
+
+1. to given room to a higher-ranked peer the router is connected to, as part of
+   the procedure to [upgrade connection slots](#slots-upgrade),
+2. or because this peer was reported as [errored](#errored-transition) by a
+   reactor interacting with this peer.
+
+This peer is a [connected peer](#connected-peer).
+`Evict` is the first sub-state of the procedure that should lead the node to
+[disconnect](#disconnected-transition) from a peer.
+
+### EvictNext transition
+
+This transition returns a peer to the router to evict.
+
+The state transition is performed whenever the peer manager has scheduled the
+eviction of a peer, i.e., whenever there is a peer on `Evict` sub-state.
+The peer to evict must be a peer in the `Connected` state.
+
+The peer to evict is randomly picked from the possible multiple peers with
+eviction scheduled.
+
+> This transition is invoked when the next to evict routine is notified by
+> another routine.
+> In some cases, the transition is processed when no peer should be evicted. In
+> this case, if the connections slots are not full, or there are enough peers
+> in the `Evicting` state so to respect the `MaxConnected` parameter, the
+> transition is not taken.
+> Otherwise, the peer with the lowest rank is evicted. This should not occur,
+> from comments in the code, but this is something to check.
+
+### Evicting peer
+
+A peer whose eviction is in progress.
+A peer transitions to this sub-state when it is returned to the router by the
+[next peer to evict](#evictnext-transition) transition.
+
+This peer is still a [connected peer](#connected-peer).
+`Evicting` is the second and last sub-state of the procedure for
+[disconnecting](#disconnected-transition) from a peer.
```

Comment:
How does it work when the peer evicts us? (maybe I've missed this in the text above)

Reply:
Good point also. This is a question that I had, put in a different way: how the remote peer knows that we rejected a connection with it?

When we evict a peer, we close the queue used to send messages to that peer. This causes the send routine for that peer to halt without errors. When either sending or receiving routines with a peer halts, we close the connection with that peer and set it as `Disconnected`.

So, if a peer evict us, we should observe an error in the receive routine with this peer, which will halt. When it halts, the associated send routine will also be stopped, and we set the peer as `Disconnected` as well.

Notice that this refers to the `Router` implementation. I have to add some details about its operation.

Reply:
Documented this mechanism in the router's spec.